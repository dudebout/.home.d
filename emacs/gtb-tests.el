;;; gtb-tests.el --- Tests for gtb.el

;;; Commentary:

;;; Code:

(require 'ert)
(require 'gtb)

(defun gtb-has-bucket-property (bucket-name)
  "Return t if the entry at point has the bucket property BUCKET-NAME."
  (equal bucket-name (org-entry-get (point) "bucket" t)))

(defun gtb-test-run ()
  "Run a test `gtb` invocation against the test buffer."
  (find-file "./gtb-tests.org")
  (gtb
   '(("bucket a"
      ((project1 . 1) (project2 . 9))
      (lambda () (gtb-has-bucket-property "a")))
     ("bucket b"
      project1
      (lambda () (gtb-has-bucket-property "b")))
     ("bucket c"
      overhead
      (lambda () (gtb-has-bucket-property "c"))
      t))))

(ert-deftest end-to-end-test ()
  (gtb-test-run)
  (let ((expected (progn
                    (find-file "./gtb-tests.golden")
                    (buffer-string)))
        (got (with-current-buffer (gtb-buffer)
               (buffer-string))))
    (should (equal got expected))))

;;; gtb-tests.el ends here
