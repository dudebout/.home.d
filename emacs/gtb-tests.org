#+PROPERTY: bucket_ALL a b
* category I
** context 1
:PROPERTIES:
:bucket:   a
:END:
*** task A
:LOGBOOK:
CLOCK: [2018-02-23 Fri 12:00]--[2018-02-23 Fri 13:00] =>  1:00
:END:
*** task B
:LOGBOOK:
CLOCK: [2018-02-23 Fri 13:00]--[2018-02-23 Fri 14:00] =>  1:00
:END:
** context 2
:PROPERTIES:
:bucket:   b
:END:
*** task A
:PROPERTIES:
:bucket:   a
:END:
:LOGBOOK:
CLOCK: [2018-02-23 Fri 14:00]--[2018-02-23 Fri 15:00] =>  1:00
:END:
*** task B
:LOGBOOK:
CLOCK: [2018-02-23 Fri 15:00]--[2018-02-23 Fri 16:00] =>  1:00
:END:
* category II
** context 1
:PROPERTIES:
:bucket:   b
:END:
*** task A
:LOGBOOK:
CLOCK: [2018-02-23 Fri 15:00]--[2018-02-23 Fri 16:00] =>  1:00
:END:
*** task B
:PROPERTIES:
:bucket:   a
:END:
:LOGBOOK:
CLOCK: [2018-02-23 Fri 16:00]--[2018-02-23 Fri 17:00] =>  1:00
:END:
*** task C
:PROPERTIES:
:bucket:   c
:END:
:LOGBOOK:
CLOCK: [2018-02-23 Fri 17:00]--[2018-02-23 Fri 18:00] =>  1:00
:END:
*** task D
:PROPERTIES:
:bucket:   d
:END:
:LOGBOOK:
CLOCK: [2018-02-23 Fri 18:00]--[2018-02-23 Fri 19:00] =>  1:00
:END:
*** task E
:LOGBOOK:
CLOCK: [2018-02-23 Fri 18:00]--[2018-02-23 Fri 18:00] =>  0:00
:END:
