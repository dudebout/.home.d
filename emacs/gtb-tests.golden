* ERROR multi-bucketed tasks
+ category II -> context 1 -> task C
* ERROR non-bucketed tasks
+ category II -> context 1 -> task D
+ category II -> context 1 -> task E
* billed
+ project1 :: 40.00%
+ project2 :: 60.00%
* billable time: 6:00
|------------------------------+-------+-------+-------+------|
| bucket a                     | 66.7% |       |       |      |
| project1: 0.1, project2: 0.9 |       |       |       |      |
|------------------------------+-------+-------+-------+------|
| category I                   | 50.0% |       |       |      |
|     context 1                |       | 33.3% |       |      |
|         task A               |       |       | 16.7% | 1:00 |
|         task B               |       |       | 16.7% | 1:00 |
|     context 2                |       | 16.7% |       |      |
|         task A               |       |       | 16.7% | 1:00 |
| category II                  | 16.7% |       |       |      |
|     context 1                |       | 16.7% |       |      |
|         task B               |       |       | 16.7% | 1:00 |
|------------------------------+-------+-------+-------+------|
| bucket b                     | 33.3% |       |       |      |
| project1                     |       |       |       |      |
|------------------------------+-------+-------+-------+------|
| category I                   | 16.7% |       |       |      |
|     context 2                |       | 16.7% |       |      |
|         task B               |       |       | 16.7% | 1:00 |
| category II                  | 16.7% |       |       |      |
|     context 1                |       | 16.7% |       |      |
|         task A               |       |       | 16.7% | 1:00 |
|------------------------------+-------+-------+-------+------|
* non-billable time: 2:00
